src/main/java/org/folio/service/UserService.java

```diff
@@ -0,0 +1,161 @@
+import org.pac4j.saml.credentials.SAML2Credentials;
+import org.pac4j.vertx.VertxWebContext;
+
+import javax.ws.rs.core.UriBuilder;
+import java.util.Collections;
+import java.util.List;
+
+import static org.folio.util.UserFields.*;
+
+public class UserService {
+
+  private static final Logger log = LogManager.getLogger(UserService.class);
+
+  private static final String USER_TENANT_ENDPOINT = "/user-tenants";
+  private static final String USER_TENANT_GET_ERROR = "Error getting user-tenant by %s = %s: %s";
+  private static final String MULTIPLE_MATCHING_USERS = "Multiple matching users by {} = {}";
+  private static final String USER_TENANT_DOES_NOT_HAVE_A_TENANT_ID_ERROR = "The matching user-tenant record founded by %s = %s does not have a tenant id";
+
+  public static class UserErrorException extends RuntimeException {
+    public UserErrorException(String message) {
+      super(message);
+    }
+  }
+
+  public static class UserFetchException extends RuntimeException {
+    public UserFetchException(String message) {
+      super(message);
+    }
+  }
+
+  public Future<JsonObject> getUser(WebClient webClient, SamlConfiguration configuration,
+                                    VertxWebContext webContext, SAML2Client client, SessionStore sessionStore,
+                                    OkapiHeaders parsedHeaders) {
+    String userPropertyName =
+      configuration.getUserProperty() == null ? EXTERNAL_SYSTEM_ID : configuration.getUserProperty();
+    var credentialsOptional = client.getCredentials(webContext, sessionStore);
+    var credentials =
+      (SAML2Credentials) credentialsOptional.orElseThrow(() -> new NullPointerException("Saml credentials was null"));
+
+    String samlAttributeValue =
+      getSamlAttributeValue(configuration.getSamlAttribute(), credentials.getUserProfile());
+    String usersCql = getCqlUserQuery(userPropertyName, samlAttributeValue);
+    String userQuery = UriBuilder.fromPath("/users").queryParam("query", usersCql).build().toString();
+
+    return handleGetUser(userPropertyName, samlAttributeValue, webClient, parsedHeaders)
+      .compose(tenantId -> webClient.getAbs(parsedHeaders.getUrl() + userQuery)
+        .putHeader(XOkapiHeaders.TOKEN, parsedHeaders.getToken())
+        .putHeader(XOkapiHeaders.URL, parsedHeaders.getUrl())
+        .putHeader(XOkapiHeaders.TENANT, tenantId)
+        .expect(ResponsePredicate.SC_OK)
+        .expect(ResponsePredicate.JSON)
+        .send()
+        .map(res -> {
+          JsonArray users = res.bodyAsJsonObject().getJsonArray("users");
+          if (users.isEmpty()) {
+            String message = "No user found by " + userPropertyName + " == " + samlAttributeValue;
+            throw new UserErrorException(message);
+          }
+          return users.getJsonObject(0);
+        }));
+  }
+
+  private Future<String> handleGetUser(String userPropertyName, String value, WebClient webClient, OkapiHeaders okapiHeaders) {
```

Comment:
in this method we actually are trying to find actula tenantId where user was created in case if ECS mode enabled.
Propose to rename to: getUserTenantId.
Also although this is a private method, I think java doc would be useful where need to specify this thing:
if ECS mode disabled - this table user tenant would be always empty and we will user orignal tenantId from request